packages/agents/sdk/src/sdkBase.ts

```diff
@@ -54,87 +48,162 @@ export class NxtpSdkBase extends NxtpSdkShared {
     }
-    const { origin, relayerFee, destination, to, asset, amount, slippage, receiveLocal } = args;
-
-    // Substitute default values as needed.
-    const callData = args.callData ?? "0x";
-    const delegate = args.delegate ?? to; // Default to using the user's signer address as the delegate.
-
-    // TODO: Calculate estimate for relayer fee and include it in the call params.
+    // Get the origin chain ID.
+    const chainId = await this.getChainId(origin);
+    // Get Connext address for origin.
+    const connextContractAddress = await this.getDeploymentAddress(origin, "connext");
 
-    // Validate XCall arguments.
-    if (asset === constants.AddressZero && amount !== "0") {
+    /// MARK - Validate arguments.
+    if (asset === constants.AddressZero && _amount !== "0") {
       // TODO: Custom error.
       throw new Error("Transacting asset specified was address zero; native assets are not supported!");
     }
-    // TODO: Should additionally make sure the asset is set to address(0) if the amount is 0.
-
-    // TODO: Use ajv validator:
-    // const validateInput = ajv.compile(XTransferSchema);
-    // const validInput = validateInput(params);
-    // if (!validInput) {
-    //   const msg = validateInput.errors?.map((err: any) => `${err.instancePath} - ${err.message}`).join(",");
-    //   throw new ParamsInvalid({
-    //     paramsError: msg,
-    //     params,
-    //   });
-    // }
 
-    const ConnextContractAddress = (await this.getConnext(origin)).address;
-
-    // Get the origin chain ID.
-    let chainId = this.config.chains[origin].chainId;
-    if (!chainId) {
-      chainId = await getChainIdFromDomain(origin, this.chainData);
+    // Sanitize arguments. Substitute default values as needed.
+    const amount = BigNumber.from(_amount);
+    let callData = args.callData ?? "0x";
+    const delegate = args.delegate ?? to; // Default to using the user's signer address as the delegate.
+    const relayerFee = BigNumber.from(_relayerFee ?? "0");
+    // TODO: Calculate estimate for relayer fee and include it in the call params.
+    // TODO: Should additionally make sure the asset is set to address(0) if the amount is 0.
+    // TODO: Use ajv validator.
+
+    if (unwrapNativeOnDestination) {
+      // Sanity check: We'll need to overwrite the callData with an unwrapping call. If a `callData` argument
+      // is also specified, we throw to ensure the user's callData isn't being overwritten unexpectedly.
+      // TODO: Implement an xReceive Multisend call to have the option to combine the unwrapping AND the user's
+      // original callData!
+      if (callData !== "0x") {
+        throw new CannotUnwrapOnDestination("`callData` argument must be empty!" + ` callData specified: ${callData}`);
+      }
+      // NOTE: We don't need to check on-chain to ensure destination Unwrapper supports unwrapping the
+      // target native token; if it didn't, it wouldn't be deployed and therefore wouldn't be configured.
```

Comment:
What happens if someone calls this with:
```
asset: something other than ETH
wrapNativeOnOrigin: true
unwrapNativeOnDestination: true
```

Will probably revert on the call? Can we check that the asset is actually wrappable and throw a good error?

Reply:
For `wrapNativeOnOrigin`:
At some point the consumer will call eth_estimateGas with the tx we've made and it will revert. But I've thought about this - it's kind of tough because we'd need to have a way to evaluate whether it *is* in fact *a* wrapper contract... we could of course have *our* set of "official" wrapper contracts (like as part of chainData or something) and just check if this is one of those assets... ultimately the only really sustainable solution here adds on to our responsibility of custodially managing "approved/acceptable" assets. I just feel like that responsibility should come from consumer.

Okay but one thing we could do I suppose is, if we have an injected provider, call eth_estimateGas ourselves on the "deposit" and "approve" method calls and make sure those pass, but that kind of feels like it's outside the scope of this method. After all, why aren't we calling eth_estimateGas on `xcall` itself inside this method? We just take for granted that the configured connext contract address in the SDK will be the official connext address.

Reply:
For `unwrapNativeOnDestination`:
There's a bunch of diff failure paths here.

1. If the user sends a wrapped token that doesn't match the wrapper contract on destination (e.g. sending WETH to Polygon and then wanting it to unwrapNativeOnDestination won't work, because Unwrapper will point to the WMATIC contract), we'll need to have a fallback route for handling that in the contract itself of course. But can we prevent that from happening here in the SDK, at the outset? Again, we could keep a sort of custodial DB/mapping here that will tell us  what the sending asset *should* be if we want to unwrap on destination (i.e. "tell me what the wrapped polygon native token address is on origin - is it the token we're sending?").

2. We use the `to` address as the "intended recipient" argument for Unwrapper, since we have to replace `to` with the Unwrapper's address here. If the user sends xcall that's missing their `to` address... we effectively prevent that here in the SDK by checking to make sure that a valid `to` value is provided (as in, the recipient can't be address(0), or we throw an error - I just added a check for this!).

Reply:
Actually for (1) I think we can check this. We should be able to do it this way:
1. get the canonicalId of the sending asset
2. get the adopted asset on destination
3. get the Unwrapper.WRAPPER property (there's a getter for that)
4. ensure Unwrapper.WRAPPER == adopted on destination. 

Right?

Reply:
Looks right for (1).

For (2), how would someone get onto the `unwrapNativeOnDestination` flow without using SDK? They'd have to encode it in `callData` and at that point the `to` address *should* be the Unwrapper.

Reply:
From internal chat:

for `wrapNativeOnOrigin`, isn't there only one supported WETH contract per chain? feels like we would be able to assert that `asset` is the WETH contract for this origin domain when you set `wrapNativeOnOrigin=true`

Reply:
re (2) see [this comment](https://github.com/connext/monorepo/pull/2992#discussion_r1067461552) i just added. we already have the concept of "oh shit i need to make adjustments on the destination chain" eoa in the agent, so i think we can lean on that if it becomes irretrievable for some reason

for 3 -- it seems like it should be a wrapper <> receiver pairing. each receiver contract has a designated wrapper (weth8, weth9, etc.) and the sdk determines which one to use based on the version it wrapped?

Reply:
actually an additional thought on (3) -- why does it matter? can we just try to send a `withdraw` via `.call` then if it fails we can assume the `asset` isn't weth? if it does, great. there's the obvious footguns of `xcall`ing a bad asset, but i dont know if it should matter what weth version it is